### ISSUE TITLE
NEATTranscriberBain not optimising

### ISSUE THREAD

**Issue by davypaperboy** [OliverColeman/ahni#7]
I am experiencing issues in getting the following simple fitness function to optimise when using the NEATTranscriberBain. The code optimises correctly when used with the ANJI library.
 
Here is the source for the base fitness function class I am using:

https://github.com/davypaperboy/ahni/blob/feature/noveltylogic/src/com/ojcoleman/ahni/experiments/logicgates/LogicGateFitness.java

with the subclass for XOR:

https://github.com/davypaperboy/ahni/blob/feature/noveltylogic/src/com/ojcoleman/ahni/experiments/logicgates/XORFitness.java

and the property file:

https://github.com/davypaperboy/ahni/blob/feature/noveltylogic/properties/noveltylogic.properties

thanks,

Davy

**Comment by OliverColeman** [comment]
Yep, there's definitely something not right here. I tried the AnjiNetTranscriber (which produces NNs built from the AnjiNet class) and the HyperNEATTranscriberBain transcriber (producing NNs built on the BainNN class) and both of these typically produce solutions within 150 to 200 generations, so the Bain NN seems fine, and a direct encoding (using AnjiNet) seems fine. There's something awry about the way NEATTranscriberBain is producing BainNNs.

I disabled multi-threading in my tests (fitness.max_threads=1) so I don't think it's a threading thing. I had a look at the chromosome to NN mapping shown in some of the "best_performing-GEN-ID.txt" files and that seems fine, though that only shows a mapping for a champion individual and only after the evaluations have been performed for a generation so there could be differences in the transcriptions performed for evaluations.

Next things to check are:
* BainNN.stepsPerStep is set correctly (how many times do we call step() (which calls SynapseCollection.step() and NeuronCollection.step()) for a call to next());
* the input (stimulus) to the network is maintained while the network is being step()'ed;
* transcriptions for evaluations are being performed correctly.

**Comment by OliverColeman** [comment]
Found the problem. `BainNN#setStepsPerStepForNonLayeredFF()` was erroneously classifying networks as cyclic/recurrent (and thus not correctly setting stepsPerStep). I've just pushed an update that fixes this. 

To help isolate the problem I set-up a XOR experiment using `com.ojcoleman.ahni.experiments.TestTargetFitnessFunctionNEAT` with `fitness.function.test.type=parity`
The full experiment set-up is in noveltylogic.properties.
